rust_src/src/keymap.rs

```diff
@@ -317,6 +316,92 @@ pub fn map_keymap_lisp(function: LispObject, keymap: LispObject, sort_first: boo
     LispObject::constant_nil()
 }
 
+/// Call FUN for every binding in MAP and stop at (and return) the parent.
+/// FUN is called with 4 arguments: FUN (KEY, BINDING, ARGS, DATA).
+#[no_mangle]
+pub extern "C" fn map_keymap_internal(
+    map: Lisp_Object,
+    fun: map_keymap_function_t,
+    args: Lisp_Object,
+    data: *const c_void,
+) -> Lisp_Object {
+    let map = LispObject::from_raw(map);
+    let mut tail = match map.as_cons() {
+        None => LispObject::constant_nil(),
+        Some(cons) => {
+            let (car, cdr) = cons.as_tuple();
+            if car.eq_raw(Qkeymap) {
+                cdr
+            } else {
+                map
+            }
+        }
+    };
+
+    for tail_cons in tail.iter_tails_safe() {
+        let binding = tail_cons.car();
+        if binding.eq_raw(Qkeymap) {
+            break;
+        } else {
+            // An embedded parent.
+            if keymapp(binding) {
+                break;
+            }
+
+            if let Some(binding_cons) = binding.as_cons() {
+                let (car, cdr) = binding_cons.as_tuple();
+                unsafe { map_keymap_item(fun, args, car.to_raw(), cdr.to_raw(), data) };
+            } else if binding.is_vector() {
+                if let Some(binding_vec) = binding.as_vectorlike() {
+                    for c in 0..binding_vec.pseudovector_size() {
+                        let character = LispObject::from_natnum(c);
+                        unsafe {
+                            map_keymap_item(
+                                fun,
+                                args,
+                                character.to_raw(),
+                                aref(binding, c).to_raw(),
+                                data,
+                            )
+                        };
+                    }
+                }
+            } else if binding.is_char_table() {
+                unsafe {
+                    let ptr = fun as *const ();
+                    let funcptr = mem::transmute(ptr);
+
+                    map_char_table(
+                        map_keymap_char_table_item,
+                        Qnil,
+                        binding.to_raw(),
+                        make_save_funcptr_ptr_obj(funcptr, data, args),
+                    );
+                }
+            }
+        }
+
+        tail = tail_cons.cdr();
```

Comment:
Why are you assigning to tail here? This is what the Iterator is doing for you.

Reply:
Ah. I ported this line without thinking that it was not needed anymore. Will fix.

Reply:
I double checked the code and the reason for `tail = tail_cons.cdr()` is because `tail` is where the parent keymap is stored as a return for the function.

Reply:
So just to reiterate this. The `tail = tail_cons.cdr()` cannot be removed as it is needed for the function to work properly.

Maybe to not make it confusing I can move this line further up after the for loop line. Hmm.